implementations/rust/ockam/ockam_identity/src/secure_channel/encryptor_worker.rs

```diff
@@ -110,14 +206,76 @@ impl EncryptorWorker {
             }
         };
 
+        info!(
+            "Sending credentials refresh for {}",
+            self.addresses.encryptor
+        );
+
         // Send the message to the decryptor on the other side
         ctx.send_from_address(
             self.remote_route.clone(),
-            encrypted_payload,
+            msg,
             self.addresses.encryptor.clone(),
         )
         .await?;
 
+        self.schedule_credentials_refresh(ctx, false).await?;
+
+        Ok(())
+    }
+
+    async fn send_close_channel(&mut self, ctx: &Context) -> Result<()> {
+        let msg = SecureChannelMessage::Close;
+
+        // Encrypt the message
+        let msg = self.encryptor.encrypt(&minicbor::to_vec(&msg)?).await?;
+
+        // Send the message to the decryptor on the other side
+        ctx.send_from_address(
+            self.remote_route.clone(),
+            msg,
+            self.addresses.encryptor.clone(),
+        )
+        .await?;
+
+        Ok(())
+    }
+
+    async fn schedule_credentials_refresh(&mut self, ctx: &Context, is_retry: bool) -> Result<()> {
+        if let Some(min_credential_expiration) = self.min_credential_expiration {
+            self.credential_refresh_event = None;
+
+            let mut credential_refresh_event =
+                DelayedEvent::create(ctx, self.addresses.encryptor_internal.clone(), ()).await?;
+
+            let now = now()?;
+
+            let duration = if min_credential_expiration < now + self.refresh_credential_time_gap {
+                self.min_refresh_credential_interval
+            } else {
+                Duration::from_secs(
+                    *(min_credential_expiration
```

Comment:
It seems that we can compute this quantity once and use it in the `if` condition and in that branch. 
Otherwise I don't fully understand the logic. If `min_credential_expiration < now + self.refresh_credential_time_gap`, doesn't that mean that we are past the time for refreshing the credential and we should then refresh now and not in `min_refresh_credential_interval`?

I also don't get the meaning of `is_retry`. What are we retrying with this boolean?

Reply:
This quantity is used only here and only in the branch that guarantees we don't go into negatives

Reply:
Yes, I think we can put 0 in the other branch here. However, I would like to avoid situation when the credential retriever returns an error really fast, and we end up calling it very many times in a loop. In this case is_retry is true and we add the delay before calling retriever again